src/Icicle/Core/Exp/Simp.hs

```diff
@@ -29,40 +29,50 @@ import qualified Data.Set                       as Set
 --   * ...something exciting???
 --
 simp :: Ord n => a -> (C.Exp a n -> Bool) -> C.Exp a n -> Fresh n (C.Exp a n)
-simp a_fresh isValue = anormal a_fresh . deadX . fixp (50 :: Int) (simpX a_fresh isValue)
+simp a_fresh isValue = anormal a_fresh . deadX . fixp (simpX a_fresh isValue)
  where
-  fixp 0 f = f
-  fixp n f = f . fixp (n-1) f
+  fixp f x
+   | Just x' <- f x = fixp f x'
+   | otherwise      = x
 
 
-simpX :: Ord n => a -> (C.Exp a n -> Bool) -> C.Exp a n -> C.Exp a n
+simpX :: Ord n => a -> (C.Exp a n -> Bool) -> C.Exp a n -> Maybe (C.Exp a n)
 simpX a_fresh isValue = go . beta
   where
     beta  = B.beta isValue
     go xx = case xx of
       -- * constant folding for some primitives
       XApp a p q
-        | p' <- go p
-        , q' <- go q
-        , Just (prim, as) <- takePrimApps (XApp a p' q')
-        , Just args       <- mapM (takeValue . go) as
-        -> fromMaybe (XApp a p' q') (simpP a_fresh prim args)
+        | mp <- go p
+        , mq <- go q
+        , p' <- fromMaybe p mp
+        , q' <- fromMaybe q mq
+        , x' <- XApp a p' q'
+        , Just (prim, as) <- takePrimApps x'
+        , Just args       <- mapM (takeValue . just go) as
+        -> case simpP a_fresh prim args of
+            Just x''
+             -> return x''
+            Nothing
+             | isJust mp || isJust mq
+             -> return x'
+             | otherwise
+             -> Nothing
 
       XApp a p q
-        -> XApp a (go p) (go q)
+        -> XApp a (just go p) <$> go q
 
       XLam a n t x1
-        -> XLam a n t (go x1)
+        -> XLam a n t <$> go x1
 
       XLet a n x1 x2
-        -- | not $ n `Set.member` freevars (go x2)
-        -- -> go x2
-        -- | otherwise
-        -> XLet a n (go x1) (go x2)
-
-      b@(XVar{})   -> b
-      b@(XPrim{})  -> b
-      b@(XValue{}) -> b
+        -> XLet a n (just go x1) <$> go x2
```

Comment:
what if you have ``let x = (5  + 2) in blah``,
the ``go (5 + 2)`` will return ``Just 7`` but ``go blah`` will return Nothing, so this will return Nothing?